STYLE.md

```diff
@@ -30,6 +30,20 @@ In addition, the following conventions should be followed:
   reserved 15;
   ```
 
+* Once established, service definitions should not be moved between packages.
+  The gRPC endpoint URL is inferred from package namespace, so this will break
+  client/server communication.
+
+* Any resource directly embedded in an `Any` object should not be moved between
+  packages. The type URL, which the package namespace is a part of, may be used
+  by Envoy or other API consuming code. Currently, this applies to the top-level
+  resources embedded in `DiscoveryResponse` objects, e.g. `Cluster`, `Listener`,
+  etc.
```

Comment:
I recommend we simplify the above 2 sections into one.

Once in use, no renaming of any kind is allowed to avoid breaking build or production. Even a simple API definition, such as a field name, can be referred by a configuration file stored in a source control system. A renaming would render an existing production configuration invalid.

To prevent people from using an unimplemented feature, clear mark such API elements as UNIMPLEMENTED.

Reply:
I don't think we need to be this strict today though; we can allow movement of leaf protos that aren't service definitions or embedded in `DiscoveryResponse` today. Where this will start to get crazy is if we allow use of `Any` for arbitrary filter opaque config, instead of `Struct` used today (see #1680 and https://blog.envoyproxy.io/dynamic-extensibility-and-protocol-buffers-dcd0bf0b8801). If we do allow that, the entire filter subtree would also need to freeze.

I'm open to just banning package renames if it's felt that they are too dangerous based on history and potential increased use of `Any`. I just want to know that we're doing this as a belts-and-braces thing rather than out of necessity.

Reply:
Any renaming would cause build breaks. It would not be acceptable for large codebase, which will take long time to update the code to fix the break.

We should follow what k8s does, you first create v1alpha, then graduate to v1. It is equivalent to a renaming, but the semantics is more clear. You will have 3+ months to finish the transition instead of instant build break.

Let's prohibit renaming first. We can always relax the rule or grant exception later.

Reply:
OK, so the concern here is really code breakage, which we have considered second-class historically vs. client/server incompatibility due to wire format violations or endpoint changes. I think this is fair as we scale, I think exceptions in the near term might continue to be made while we are small. We're not planning on using  `/vN` as previously discussed any time soon, since we are sticking with v2 as the API major version for some time (with additions and deprecations according to the breaking change policy).